File: THCS/2023_2024/QuangNinhA_2324/Bai2/bina.cpp
#include <iostream>
#include <fstream>
#include <string>

using namespace std;

int main() {
    ifstream input("bina.inp");
    ofstream output("bina.out");

    int n, k;
    input >> n >> k;
    string s;
    input >> s;

    int changes = 0;
    string result = s;

    for (int i = 0; i < n;) {
        int count = 1; // Đếm số bit giống nhau liên tiếp

        // Đếm số bit giống nhau liên tiếp
        while (i + 1 < n && result[i] == result[i + 1]) {
            count++;
            i++;
        }

        // Nếu số bit giống nhau liên tiếp lớn hơn k, cần sửa đổi
        if (count > k) {
            // Cần sửa đổi bit
            for (int j = k; j < count; j++) {
                // Chọn bit cần sửa
                result[i - j + 1] = (result[i - j + 1] == '1') ? '0' : '1'; // Lật bit
                changes++;
            }
        }

        i++; // Tiến tới ký tự tiếp theo
    }

    // Ghi kết quả ra tệp
    output << changes << endl;
    output << result << endl;

    return 0;
}

File: THCS/2023_2024/QuangNinhA_2324/Bai2/bina.py
def solve():
    # Đọc dữ liệu từ file
    with open("bina.inp", "r") as infile:
        n, k = map(int, infile.readline().strip().split())
        s = infile.readline().strip()

    # Biến để đếm số thao tác và lưu chuỗi sửa đổi
    operations = 0
    result = list(s)  # Chuyển đổi chuỗi thành danh sách để có thể sửa đổi

    # Duyệt qua chuỗi để kiểm tra và sửa
    count = 1  # Đếm số lượng bit giống nhau liên tiếp
    for i in range(1, n):
        if result[i] == result[i - 1]:  # Nếu bit giống nhau
            count += 1
        else:
            count = 1  # Đặt lại đếm nếu gặp bit khác

        # Nếu số lượng bit giống nhau vượt quá k
        if count > k:
            # Lật bit hiện tại
            operations += 1
            result[i] = '0' if result[i] == '1' else '1'  # Đảo bit
            count = 1  # Reset đếm về 1 vì vừa lật bit

    # Ghi kết quả ra file
    with open("bina.out", "w") as outfile:
        outfile.write(f"{operations}\n")  # Số thao tác tối thiểu
        outfile.write("".join(result))  # Chuỗi đã sửa đổi

if __name__ == "__main__":
    solve()
